File: kingsfamilylibrary/src/main/java/blue/koenig/kingsfamilylibrary/model/communication/ServerConnection.java
package blue.koenig.kingsfamilylibrary.model.communication;


import com.example.OnConnectionChangedListener;
import com.example.OnReceiveBytesListener;
import com.example.SocketChannelTCPClient;
import com.koenig.communication.ConnectUtils;
import com.koenig.communication.Parser;
import com.koenig.communication.messages.FamilyMessage;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Thomas on 08.01.2017.
 */
public class ServerConnection extends SocketChannelTCPClient implements OnConnectionChangedListener, OnReceiveBytesListener {
    private Logger logger = LoggerFactory.getLogger(getClass().getSimpleName());
    private List<ConnectionEventListener> connectionEventListeners;
    private String fromId;

    public ServerConnection(String fromId) {
        super(ConnectUtils.PORT, ConnectUtils.SERVER_IP);
        this.fromId = fromId;
        connectionEventListeners = new ArrayList<>();
        addOnConnectionChangedListener(this);
        addOnReceiveBytesListener(this);
    }

    public void setFromId(String fromId) {
        this.fromId = fromId;
    }

    @Override
    public boolean isConnected() {
        return super.isConnected();
    }

    @Override
    public void connect() {
        logger.trace("Trying to connect...");
        tryConnect();
    }

    @Override
    public void disconnect() {
        super.disconnect();
    }

    public void addOnConnectionEventListener(ConnectionEventListener connectionEventListener) {
        this.connectionEventListeners.add(connectionEventListener);
    }

    public void sendFamilyMessage(FamilyMessage msg) {
        msg.setFromId(fromId);
        msg.setToId(FamilyMessage.ServerId);
        super.sendMessage(msg);
        logger.info("Sent message " + msg.getName());
    }

    @Override
    public void onConnectionChanged(boolean b) {
        logger.info("ConnectionChanged: " + b);
        for (ConnectionEventListener connectionEventListener : connectionEventListeners) {
            connectionEventListener.onConnectionStatusChange(b);
        }

    }

    @Override
    public void onReceiveBytes(byte[] bytes) {

        logger.info("Receive bytes: " + bytes.length);
        FamilyMessage msg = Parser.parse(ByteBuffer.wrap(bytes));

        logger.info(msg.toString());
        for (ConnectionEventListener connectionEventListener : connectionEventListeners) {
            connectionEventListener.onReceiveMessage(msg);
        }

    }


}

File: kingsfamilylibrary/src/main/java/blue/koenig/kingsfamilylibrary/model/communication/ServerConnection.kt
package blue.koenig.kingsfamilylibrary.model.communication


import com.example.OnConnectionChangedListener
import com.example.OnReceiveBytesListener
import com.example.SocketChannelTCPClient
import com.koenig.communication.ConnectUtils
import com.koenig.communication.Parser
import com.koenig.communication.messages.FamilyMessage
import io.reactivex.Observable
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

/**
 * Created by Thomas on 08.01.2017.
 */
class ServerConnection(userIdObservable: Observable<String>) : SocketChannelTCPClient(ConnectUtils.PORT, ConnectUtils.SERVER_IP), OnConnectionChangedListener, OnReceiveBytesListener {
    private val logger = LoggerFactory.getLogger(javaClass.simpleName)
    private val connectionEventListeners: MutableList<ConnectionEventListener>

    private lateinit var fromId: String

    init {
        userIdObservable.subscribe { fromId = it }
        connectionEventListeners = ArrayList()
        addOnConnectionChangedListener(this)
        addOnReceiveBytesListener(this)
    }

    override fun connect() {
        logger.trace("Trying to connect...")
        tryConnect()
    }

    fun addOnConnectionEventListener(connectionEventListener: ConnectionEventListener) {
        logger.info("AddonConnectionEventListener: " + connectionEventListener.toString())
        this.connectionEventListeners.add(connectionEventListener)
    }

    fun sendFamilyMessage(msg: FamilyMessage) {
        msg.fromId = fromId
        msg.toId = FamilyMessage.ServerId
        super.sendMessage(msg)
        logger.info("Sent message " + msg.name)
    }

    override fun onConnectionChanged(b: Boolean) {
        logger.info("ConnectionChanged: " + b)
        for (connectionEventListener in connectionEventListeners) {
            connectionEventListener.onConnectionStatusChange(b)
        }

    }

    override fun onReceiveBytes(bytes: ByteArray) {

        logger.info("Receive bytes: " + bytes.size)
        val msg: FamilyMessage?
        try {
            msg = Parser.parse(ByteBuffer.wrap(bytes))

        } catch (e: Exception) {
            logger.error("Error while parsing message: " + e.message)
            return
        }

        logger.info(msg.toString())
        for (connectionEventListener in connectionEventListeners) {
            connectionEventListener.onReceiveMessage(msg)
        }

    }


}
